include/seqan3/range/decorator/gap_decorator_anchor_set.hpp

```diff
@@ -0,0 +1,485 @@
+        // case 1: extend previous/surrounding gap already existing
+        if ((pos < this->size()) && (((value_type)(*this)[pos] == gap::GAP) || (pos > 0 && (value_type)(*this)[pos-1] == gap::GAP)))
+        {
+            it_set = anchors.lower_bound(gap_t{pos, _});
+            if (it_set == anchors.end() || (*it_set).first > pos)
+                it_set = std::prev(it_set);
+            gap_t gap{(*it_set).first, (*it_set).second + size};
+            // merge with successor
+            auto it_next = it_set;
+            ++it_next;
+            if ((*it_set) < (*std::prev(anchors.end())) && (*it_next).first <= (*it_set).first + size - 1)
+            {
+                // extend gap for *it_next, delete *(it_next+1)
+                gap.second += (*it_next).second;
+                anchors.erase(it_next);
+            }
+            anchors.erase(it_set);
+            anchors.insert(gap);
+        }
+        // case 2: create new anchor gap
+        else
+        {
+            gap_t gap{pos, size};
+            // pre: pos not in anchor set, find next lower index
+            it = anchors.find(gap_t{pos, _});
+            // add accumulated gaps from preceeding gap
+            if (it != anchors.begin() && it != anchors.end())
+                gap.second += (*--it).second;
+            anchors.insert(gap);
+        }
+        // post-processing: reverse update of succeeding gaps
+        rupdate(pos, size);
+        return true;
+    }
+
+    /*!\brief Erase gap in given iterator range (exluding it2).
+    * \returns A boolean flag indicating the success of the erasing operation.
+    *
+    * \par Complexity
+    *
+    * Average and worst case (erasure before last gap): o(k),
+    * Best case (back erasure): o(log k).
+    *
+    * \par Exceptions
+    *
+    * May throw std::set exceptions.
+    */
+    bool erase_gap(iterator const it)
+    {
+        return erase_gap(it, std::next(it));
+    }
+    //!\copydoc erase_gap(iterator const it)
+    bool erase_gap(iterator const it1, iterator const it2)
+    {
+        // TODO: assert or return false?
+        if (it1 > it2 || it2 > this->end())
+            return false;
+        //assert(it1 <= it2 && it2 <= this->end());
+        size_type pos1 = it1 - this->begin(), pos2 = it2 - this->begin();
+        anchor_set_iterator it = anchors.lower_bound(gap_t{pos1, _});
+        size_type gap_len = get_gap_length(it);
+
+        if (it == anchors.end() || (*it).first > pos1)
+            it = std::prev(it);
+        // case 1: complete gap is deleted
+        if (((*it).first == pos1) && (gap_len == pos2-pos1))
+            anchors.erase(it);
+        // case 2: gap to be deleted in tail or larger than 1 (equiv. to shift tail left, i.e. pos remains unchanged)
+        else
+        {
+            gap_t gap{(*it).first, (*it).second - pos2 + pos1};
+            // TODO: emplace better?
+            anchors.erase(it);
+            anchors.insert(gap);
```

Comment:
Not sure, how to use it correctly here. The to be inserted gap is per comparator definition identical to the old one (because the key is the unmodified position element), which means that insertion would be aborted. So, my current solution is to delete the old one, which invalidates the iterator (not usable as hint anymore) and then apply a simple insert.

Reply:
yes you still have to delete the old one, but erase returns a valid iterator the position following the last removed element with is precisely the position to insert.
```cpp
auto new_it = anchors.erase(it);
anchors.emplace_hint(new_it, gap);
```